servant-client/src/Servant/Client/HttpClient.hs

```diff
@@ -0,0 +1,142 @@
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+
+{-| http-client based client  requests executor -}
+module Servant.Client.HttpClient where
+
+
+import Prelude ()
+import Prelude.Compat
+
+import Control.Exception
+import Control.Monad
+import Control.Monad.Catch (MonadThrow, MonadCatch)
+import Data.Foldable (toList)
+import Data.Functor.Alt (Alt (..))
+
+import Control.Monad.Error.Class (MonadError(..))
+import Control.Monad.Trans.Except
+
+import GHC.Generics
+import Control.Monad.Base (MonadBase (..))
+import Control.Monad.IO.Class ()
+import Control.Monad.Reader
+import Control.Monad.Trans.Control (MonadBaseControl (..))
+import Data.ByteString.Lazy hiding (pack, filter, map, null, elem, any)
+import Data.String.Conversions (cs)
+import Data.Proxy
+import Network.HTTP.Media
+import Network.HTTP.Types
+import Network.HTTP.Client hiding (Proxy, path)
+import qualified Network.HTTP.Types.Header   as HTTP
+import Servant.API.ContentTypes
+import Servant.Client.Class
+import Servant.Common.BaseUrl
+import Servant.Common.Req
+
+import qualified Network.HTTP.Client as Client
+
+instance RunClient ClientM NoContent ( Int, ByteString, MediaType
+                                     , [HTTP.Header], Response ByteString) where
+  runRequest _ meth req = performRequest meth req
+
+instance (MimeUnrender ct a) =>
+         RunClient ClientM ct ([HTTP.Header], a) where
+  runRequest p meth req = performRequestCT p meth req
+
+instance RunClient ClientM NoContent [HTTP.Header] where
+  runRequest _ meth req = performRequestNoBody meth req
+
+data ClientEnv
+  = ClientEnv
+  { manager :: Manager
+  , baseUrl :: BaseUrl
+  }
+
+
+-- | @ClientM@ is the monad in which client functions run. Contains the
+-- 'Manager' and 'BaseUrl' used for requests in the reader environment.
+
+newtype ClientM a = ClientM { runClientM' :: ReaderT ClientEnv (ExceptT ServantError IO) a }
+                    deriving ( Functor, Applicative, Monad, MonadIO, Generic
+                             , MonadReader ClientEnv
+                             , MonadError ServantError
+                             , MonadThrow, MonadCatch
+                             )
+
+instance MonadBase IO ClientM where
+  liftBase = ClientM . liftBase
+
+instance MonadBaseControl IO ClientM where
+  type StM ClientM a = Either ServantError a
+
+  -- liftBaseWith :: (RunInBase ClientM IO -> IO a) -> ClientM a
+  liftBaseWith f = ClientM (liftBaseWith (\g -> f (g . runClientM')))
+
+  -- restoreM :: StM ClientM a -> ClientM a
+  restoreM st = ClientM (restoreM st)
+
+-- | Try clients in order, last error is preserved.
+instance Alt ClientM where
+  a <!> b = a `catchError` \_ -> b
+
+runClientM :: ClientM a -> ClientEnv -> IO (Either ServantError a)
+runClientM cm env = runExceptT $ (flip runReaderT env) $ runClientM' cm
+
+
+performRequest :: Method -> Req
+               -> ClientM ( Int, ByteString, MediaType
```

Comment:
I'd make a record for this 5-tuple.

Also `Response` ties the lib to http-client still, does one need anything else than `requestBody`?

Reply:
Agreed on both counts here.

Reply:
I left mots of the existing code untouched. This should probably be part of another PR

Reply:
I commented otherwise, but the curren t `Servant.Client` module shouldn't depend on `HttpClient`, thus this change is mandatory.

Reply:
OK. I was expecting to provide this feature in such a way that existing code would stay compatible without any modification. If you want to sever dependency between `Client` and `HttpClient`, which indeed is the ultimate goal of #800 then I think it would make sense to also implement the decoupling of packages.